File: programmers/LV2/두 원 사이의 정수쌍.swift
import Foundation

func minmaxY(_ r1: Int, _ r2: Int, _ x: Int) -> Int64 {
    var min: Int64 = 0
    if x < r1 {
        let minDouble: Double = sqrt(Double(r1 * r1) - Double(x * x)) 
        min = Double(Int64(minDouble)) == minDouble ? Int64(minDouble) : Int64(minDouble) + 1
    }

    let maxDouble: Double = sqrt(Double(r2 * r2) - Double(x * x)) 
    let max = Int64(maxDouble)
    
    return max - min + 1
}

func solution(_ r1: Int, _ r2: Int) -> Int64 {
    var answer: Int64 = 0
    
    for i in 0...r2 {
        answer += minmaxY(r1, r2, i)
    }
    answer *= 4
    answer -= minmaxY(r1, r2, 0) * 4
    
    return answer
}

File: programmers/LV2/두 원 사이의 정수쌍.cpp
#include <string>
#include <vector>
#include <cmath>

using namespace std;

// c++ 문제 풀때 단순곱에서 overflow 발생하는 case 조심하기!
long long minmaxY(long long r1, long long r2, long long x) {
    long long min = 0, max = 0;
    
    if(x < r1) {
        double minDouble = sqrt((double)(r1 * r1 - x * x));
        long long minLong = (long long)minDouble;
        min = (double)minLong == minDouble ? minLong : minLong + 1;
    }
    
    double maxDouble = sqrt((double)(r2 * r2 - x * x));
    max = (long long)maxDouble;
    
    return max - min + 1;
}

long long solution(int r1, int r2) {
    long long answer = 0;
    
    for(int i = 0; i <= r2; i++) {
        answer += minmaxY(r1, r2, i);
    }
    answer *= 4;
    answer -= minmaxY(r1, r2, 0) * 4;
    
    return answer;
}